teos/src/tx_index.rs

```diff
@@ -0,0 +1,460 @@
+    fn _init(&mut self, last_n_blocks: &[ValidatedBlock]) {
+        for block in last_n_blocks.iter().rev() {
+            if let Some(prev_block_hash) = self.blocks().back() {
+                if block.header.prev_blockhash != *prev_block_hash {
+                    panic!("last_n_blocks contains unchained blocks");
+                }
+            };
+
+            let map = block
+                .txdata
+                .iter()
+                .map(|tx| {
+                    (
+                        K::from_txid(tx.txid()),
+                        match V::get_type() {
+                            Type::Transaction => V::from_data(Data::Transaction(tx.clone())),
+                            Type::BlockHash => {
+                                V::from_data(Data::BlockHash(block.header.block_hash()))
+                            }
+                        },
+                    )
+                })
+                .collect();
+
+            self.update(block.header, &map);
+        }
+    }
+
+    /// Updates the index by adding data from a new block. Removes the oldest block if the index is full afterwards.
+    fn update(&mut self, block_header: BlockHeader, data: &HashMap<K, V>) {
+        self.blocks_mut().push_back(block_header.block_hash());
+
+        let ks = data
+            .iter()
+            .map(|(k, v)| {
+                self.index_mut().insert(*k, v.clone());
+                *k
+            })
+            .collect();
+
+        self.tx_in_block_mut().insert(block_header.block_hash(), ks);
+
+        if self.is_full() {
+            // Avoid logging during bootstrap
+            log::info!("New block added to index: {}", block_header.block_hash());
+            self.update_tip();
+            self.remove_oldest_block();
+        }
+    }
+
+    /// Fixes the index by removing disconnected data.
+    fn remove_disconnected_block(&mut self, header: &BlockHeader) {
+        if let Some(ks) = self.tx_in_block_mut().remove(&header.block_hash()) {
+            self.index_mut().retain(|k, _| !ks.contains(k));
+
+            // Blocks should be disconnected from last backwards. Log if that's not the case so we can revisit this and fix it.
+            if let Some(h) = self.blocks_mut().pop_back() {
+                if h != header.block_hash() {
+                    log::error!("Disconnected block does not match the oldest block stored in the LocatorCache ({} != {})", header.block_hash(), h);
+                }
+            }
+        } else {
+            log::warn!("The index is already empty");
+        }
+    }
+
+    /// Removes the oldest block from the index.
+    /// This removes data from `self.blocks`, `self.tx_in_block` and `self.index`.
+    fn remove_oldest_block(&mut self) {
+        let h = self.blocks_mut().pop_front().unwrap();
+        let ks = self.tx_in_block_mut().remove(&h).unwrap();
+        self.index_mut().retain(|k, _| !ks.contains(k));
```

Comment:
The retain here is not removing the last block per see, but the data (e.g. transactions) that was included in that block. Therefore it may (and most likely would in production) imply several calls to remove.

Reply:
Oooh, my bad, I thought this is line was removing a specific block from the (block -> tx) map (thus one element would be removed, but that's actually L#227).